File: php/src/detectors/RfiDetector.php
<?php

namespace FireWTWall\Detectors;

use FireWTWall\Request;

/**
 * Detects Remote File Inclusion (RFI) attempts.
 *
 * Only activates when the parameter name is one of a known set of file/path
 * carrying names.  Checks each value for remote URL schemes and dangerous
 * pseudo-protocols that could cause the server to include a remote resource.
 *
 * Requires PHP >= 8.0
 */
class RfiDetector
{
	/** Parameter names commonly used to pass file or include paths */
	private const FILE_PARAMS = [
		'page', 'file', 'include', 'require', 'template', 'view', 'document',
		'folder', 'root', 'path', 'pg', 'style', 'pdf', 'layout', 'conf',
		'config', 'inc', 'mod', 'module', 'load', 'show',
	];

	private static array $rules = [
		['id' => 'rfi-http',       'pattern' => '/^https?:\/\//i',                 'severity' => 'critical'],
		['id' => 'rfi-ftp',        'pattern' => '/^ftp:\/\//i',                    'severity' => 'critical'],
		['id' => 'rfi-smb',        'pattern' => '/^\\\\\\\\/i',                    'severity' => 'critical'],
		['id' => 'rfi-expect',     'pattern' => '/^expect:\/\//i',                  'severity' => 'critical'],
		['id' => 'rfi-data',       'pattern' => '/^data:text\/plain;base64,/i',    'severity' => 'critical'],
		['id' => 'rfi-log-poison', 'pattern' => '/\/var\/log\/(apache|nginx|httpd|auth|syslog|mail)|\/proc\/self\/environ/i', 'severity' => 'critical'],
	];

	/**
	 * Scan the request for RFI payloads.
	 *
	 * @param  Request $request  Normalised request object
	 * @param  array   $config   WAF configuration (reserved for future use)
	 * @return array{rule:string,matched:string,source:string,severity:string}|null
	 */
	public static function scan(Request $request, array $config): ?array
	{
		$allSources = [
			'query'   => $request->getQuery(),
			'body'    => $request->getBody(),
			'cookies' => $request->getCookies(),
		];

		foreach ($allSources as $sourceLabel => $data) {
			$result = self::scanFlat($data, $sourceLabel);
			if ($result !== null) {
				return $result;
			}
		}

		return null;
	}

	// ------------------------------------------------------------------ //
	// Private helpers
	// ------------------------------------------------------------------ //

	/**
	 * Walk a (possibly nested) array and check values whose key is a known
	 * file/include parameter name.
	 */
	private static function scanFlat(array $data, string $sourceLabel, string $prefix = ''): ?array
	{
		foreach ($data as $key => $value) {
			if (is_array($value)) {
				$paramName = $prefix !== '' ? $prefix . '[' . $key . ']' : (string) $key;
				$result    = self::scanFlat($value, $sourceLabel, $paramName);
				if ($result !== null) {
					return $result;
				}
				continue;
			}

			if (!is_string($value)) {
				continue;
			}

			// Only inspect parameters whose name suggests a file/include path
			if (!in_array(strtolower((string) $key), self::FILE_PARAMS, true)) {
				continue;
			}

			foreach (self::$rules as $rule) {
				if (preg_match($rule['pattern'], $value, $m)) {
					return [
						'rule'     => $rule['id'],
						'severity' => $rule['severity'],
						'matched'  => substr($m[0], 0, 120),
						'source'   => $sourceLabel,
					];
				}
			}
		}

		return null;
	}
}

File: aspnet/src/detectors/RfiDetector.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    /// <summary>
    /// Detects Remote File Inclusion (RFI) attempts.
    /// </summary>
    public static class RfiDetector
    {
        private static readonly HashSet<string> FileParams = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
        {
            "page", "file", "include", "require", "template", "view", "document",
            "folder", "root", "path", "pg", "style", "pdf", "layout", "conf",
            "config", "inc", "mod", "module", "load", "show",
        };

        private static readonly (string Id, string Severity, Regex Pattern)[] Rules =
        {
            ("rfi-http",       "critical", new Regex(@"^https?://",                                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("rfi-ftp",        "critical", new Regex(@"^ftp://",                                                      RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("rfi-smb",        "critical", new Regex(@"^\\\\",                                                        RegexOptions.Compiled)),
            ("rfi-expect",     "critical", new Regex(@"^expect://",                                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("rfi-data",       "critical", new Regex(@"^data:text/plain;base64,",                                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("rfi-log-poison", "critical", new Regex(@"/var/log/(apache|nginx|httpd|auth|syslog|mail)|/proc/self/environ", RegexOptions.IgnoreCase | RegexOptions.Compiled)),
        };

        public static DetectorResult Scan(WafRequest request)
        {
            foreach (var source in new[] {
                ("query",   request.Query),
                ("body",    request.Form),
                ("cookies", request.Cookies)
            })
            {
                foreach (var kv in source.Item2)
                {
                    if (!FileParams.Contains(kv.Key)) continue;
                    foreach (var rule in Rules)
                    {
                        Match m = rule.Pattern.Match(kv.Value);
                        if (m.Success)
                            return new DetectorResult(rule.Id, rule.Severity,
                                m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value,
                                source.Item1);
                    }
                }
            }
            return null;
        }
    }
}
